test/search/test_flann_search.cpp

```diff
@@ -257,7 +250,7 @@ TEST (PCL, FlannSearch_knnByIndex)
 
   }
 
-  delete(flann_search);
+  delete (flann_search);
```

Comment:
You forgot to convert this variable from a pointer type.

Reply:
https://github.com/PointCloudLibrary/pcl/blob/6c1e94b6d06d66d0c08de4a18f82ff24a97332aa/test/search/test_flann_search.cpp#L243 This `nearestKSearch` is not implemented in `FlannSearch`. So I could not change it.

Reply:
Assuming you really changed the pointer to a normal variable, if it was really not implemented, then `flann_search->nearestKSearch` would also have yielded a compiler error. The problem is likely something else.

Reply:
https://github.com/PointCloudLibrary/pcl/blob/6c1e94b6d06d66d0c08de4a18f82ff24a97332aa/test/search/test_flann_search.cpp#L216
This cast to `pcl::search::Search<PointXYZ>` probably does something as the function is implemented in the parent class

Reply:
The child class inherits all the public methods from base, so having the cast to base should not matter as long as the methods are virtual, which they appear to be. Can you post here the diff with the changes you tried and the compilation error log?

Reply:
Diff
```
-  pcl::search::Search<PointXYZ>* flann_search = new pcl::search::FlannSearch<PointXYZ> (new search::FlannSearch<PointXYZ>::KdTreeIndexCreator);
+  pcl::search::FlannSearch<PointXYZ> flann_search (new search::FlannSearch<PointXYZ>::KdTreeIndexCreator)
```

Error: 

```
pcl/test/search/test_flann_search.cpp:243:101: error: no matching function for call to ‘pcl::search::FlannSearch<pcl::PointXYZ>::nearestKSearch(pcl::PointCloud<pcl::PointXYZ>&, __gnu_cxx::__alloc_traits<std::allocator<int> >::value_type&, unsigned int&, std::vector<int>&, std::vector<float>&)’
     flann_search.nearestKSearch (cloud_big,query_indices[i], no_of_neighbors, k_indices, k_distances);

```

Reply:
It was a little more involved than I expected
```diff
diff --git a/search/include/pcl/search/flann_search.h b/search/include/pcl/search/flann_search.h
index 95f61056d..6dd7e63a7 100644
--- a/search/include/pcl/search/flann_search.h
+++ b/search/include/pcl/search/flann_search.h
@@ -260,6 +260,8 @@ namespace pcl
         void
         setInputCloud (const PointCloudConstPtr& cloud, const IndicesConstPtr& indices = IndicesConstPtr ()) override;
 
+        using Search<PointT>::nearestKSearch;
+
         /** \brief Search for the k-nearest neighbors for the given query point.
           * \param[in] point the given query point
           * \param[in] k the number of neighbors to search for
diff --git a/test/search/test_flann_search.cpp b/test/search/test_flann_search.cpp
index b3f28ddee..b1ff6d3bd 100644
--- a/test/search/test_flann_search.cpp
+++ b/test/search/test_flann_search.cpp
@@ -213,9 +213,12 @@ TEST (PCL, FlannSearch_knnByIndex)
   unsigned int no_of_neighbors = 3;
 
 
-  pcl::search::Search<PointXYZ>* flann_search = new pcl::search::FlannSearch<PointXYZ> (new search::FlannSearch<PointXYZ>::KdTreeIndexCreator);
+  pcl::search::FlannSearch<PointXYZ> flann_search (
+    true,
+    search::FlannSearch<PointXYZ>::FlannIndexCreatorPtr (new search::FlannSearch<PointXYZ>::KdTreeIndexCreator)
+  );
   //FlannSearch->initSearchDS ();
-  flann_search->setInputCloud (cloud_big.makeShared ());
+  flann_search.setInputCloud (cloud_big.makeShared ());
 
   std::vector< std::vector< float > > dists;
   std::vector< std::vector< int > > indices;
@@ -224,7 +227,7 @@ TEST (PCL, FlannSearch_knnByIndex)
   {
     query_indices.push_back (int (i));
   }
-  flann_search->nearestKSearch (cloud_big, query_indices,no_of_neighbors,indices,dists);
+  flann_search.nearestKSearch (cloud_big, query_indices,no_of_neighbors,indices,dists);
 
   std::vector<int> k_indices;
   k_indices.resize (no_of_neighbors);
@@ -233,14 +236,14 @@ TEST (PCL, FlannSearch_knnByIndex)
 
   for (std::size_t i = 0; i < query_indices.size (); ++i)
   {
-    flann_search->nearestKSearch (cloud_big[2*i], no_of_neighbors, k_indices, k_distances);
+    flann_search.nearestKSearch (cloud_big[2*i], no_of_neighbors, k_indices, k_distances);
     EXPECT_EQ (k_indices.size (), indices[i].size ());
     EXPECT_EQ (k_distances.size (), dists[i].size ());
     for (std::size_t j = 0; j< no_of_neighbors; j++)
     {
       EXPECT_TRUE (k_indices[j] == indices[i][j] || k_distances[j] == dists[i][j]);
     }
-    flann_search->nearestKSearch (cloud_big,query_indices[i], no_of_neighbors, k_indices, k_distances);
+    flann_search.nearestKSearch (cloud_big, query_indices[i], no_of_neighbors, k_indices, k_distances);
     EXPECT_EQ (k_indices.size (), indices[i].size ());
     EXPECT_EQ (k_distances.size (), dists[i].size ());
     for (std::size_t j = 0; j< no_of_neighbors; j++)
```

Intuitions into what happened:
https://stackoverflow.com/questions/1896830/why-should-i-use-the-using-keyword-to-access-my-base-class-method
https://isocpp.org/wiki/faq/strange-inheritance#hiding-rule

Reply:
> Are we likely gonna change this default?

That doesn't matter, right? We're not testing the mechanism here, so the new version is actually better IMO since it tests the actual class as presented to the user.